src/lib/staged_ledger/staged_ledger.ml

```diff
@@ -1994,29 +1817,37 @@ let%test_module "test" =
       let initial_ledger = ref 0 in
       let sl = ref (Sl.create ~ledger:initial_ledger) in
       Quickcheck.test g ~trials:1000 ~f:(fun _ ->
-          let old_ledger = !(Sl.ledger !sl) in
-          let all_ts = txns (p / 2) (fun x -> (x + 1) * 100) (fun _ -> 4) in
-          let ledger_proof, diff =
-            create_and_apply sl logger (Sequence.of_list all_ts) stmt_to_work
-            |> Or_error.ok_exn
-          in
-          let fee_excess =
-            Option.value_map ~default:Currency.Fee.Signed.zero ledger_proof
-              ~f:(fun proof ->
-                let stmt = Test_input1.Ledger_proof.statement proof in
-                stmt.fee_excess )
-          in
-          (*fee_excess at the top should always be zero*)
-          assert (Currency.Fee.Signed.equal fee_excess Currency.Fee.Signed.zero) ;
-          let cb = coinbase_added diff in
-          (*At worst case number of provers coinbase should not be split more than two times*)
-          assert (cb > 0 && cb < 3) ;
-          let x =
-            List.length (Test_input1.Staged_ledger_diff.user_commands diff)
-          in
-          assert_at_least_coinbase_added x cb ;
-          let expected_value = expected_ledger x all_ts old_ledger in
-          assert (!(Sl.ledger !sl) = expected_value) )
+          Async.Thread_safe.block_on_async_exn (fun () ->
```

Comment:
Probably better to localize the block on async as tightly as possible over the deferred section

Reply:
Putting it at the highest level in the expression helps to avoid the ocaml compiler error that can happen with Async.Thread_safe.block_on_async_exn code. This works because anything inside the function passed into the thread safe block will not be considered for unused mutable constant popagation, which is the compiler optimization I've seen mess up with async blocks before.